subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/Naggers.java

```diff
@@ -0,0 +1,325 @@
+     */
+    public static void setTraceLoggingEnabled(boolean traceLoggingEnabled) {
+        Naggers.traceLoggingEnabled = traceLoggingEnabled;
+    }
+
+    static boolean isTraceLoggingEnabled() {
+        String value = System.getProperty(ORG_GRADLE_DEPRECATION_TRACE_PROPERTY_NAME);
+        if (value == null) {
+            return traceLoggingEnabled;
+        }
+        return Boolean.parseBoolean(value);
+    }
+
+    private static class DefaultBasicNagger implements BasicNagger {
+        private static final Set<String> PREVIOUS_MESSAGES = new HashSet<String>();
+        private static final LoggingFeatureUsageHandler HANDLER = new LoggingFeatureUsageHandler();
+
+        public void nagUserWith(String message) {
+            if (ENABLED.get()) {
+                LOCK.lock();
+                try {
+                    HANDLER.featureUsed(new FeatureUsage(message));
+                } finally {
+                    LOCK.unlock();
+                }
+            }
+
+        }
+
+        public void nagUserOnceWith(String message) {
+            if (ENABLED.get()) {
+                LOCK.lock();
+                try {
+                    if (PREVIOUS_MESSAGES.add(message)) {
+                        HANDLER.featureUsed(new FeatureUsage(message));
+                    }
+                } finally {
+                    LOCK.unlock();
+                }
+            }
+        }
+
+        private void reset() {
+            LOCK.lock();
+            try {
+                PREVIOUS_MESSAGES.clear();
+            } finally {
+                LOCK.unlock();
+            }
+        }
+
+        private void useLocationReporter(UsageLocationReporter reporter) {
+            LOCK.lock();
+            try {
+                HANDLER.setLocationReporter(reporter);
+            } finally {
+                LOCK.unlock();
+            }
+        }
+    }
+
+    private static class DefaultIncubationNagger implements IncubationNagger {
+
+        public void incubatingFeatureUsed(String incubatingFeature) {
+            incubatingFeatureUsed(incubatingFeature, null);
+        }
+
+        public void incubatingFeatureUsed(String incubatingFeature, String additionalWarning) {
+            String message = getIncubationMessage(incubatingFeature);
+            if (additionalWarning != null) {
+                message = message + '\n' + additionalWarning;
+            }
+            getBasicNagger().nagUserOnceWith(message);
+        }
+
+        private static String getIncubationMessage(String incubatingFeature) {
+            return String.format("%s is an incubating feature.", incubatingFeature);
+        }
+    }
+
+    private static class DefaultDeprecationNagger implements DeprecationNagger {
+        private static String deprecationMessage;
+
+        public void nagUserOfDeprecated(String thing) {
+            getBasicNagger().nagUserWith(String.format("%s %s.", thing, getDeprecationMessage()));
```

Comment:
It would be trivial to switch back to the previous behavior of only warning once per deprecation message by changing the `nagUserWith` calls in `DefaultDeprecationNagger` to `nagUserOnceWith` calls. This would keep the architectural benefits while keeping behavior unchanged. Please don't, though.